lib/vsc/utils/filetools.py

```diff
@@ -0,0 +1,201 @@
+        _log.raiseException(msg)
+    elif fail_cnt > 0:
+        _log.debug("%.2f%% of permissions/owner operations failed, ignoring that..." % (100 * fail_ratio))
+
+
+def mkdir(path, parents=False, set_gid=None, sticky=None):
+    """
+    Create a directory
+    Directory is the path to create
+
+    @param parents: create parent directories if needed (mkdir -p)
+    @param set_gid: set group ID bit, to make subdirectories and files inherit group
+    @param sticky: set the sticky bit on this directory (a.k.a. the restricted deletion flag),
+                   to avoid users can removing/renaming files in this directory
+    """
+    if not os.path.isabs(path):
+        path = os.path.abspath(path)
+
+    # exit early if path already exists
+    if not os.path.exists(path):
+        tup = (path, parents, set_gid, sticky)
+        _log.info("Creating directory %s (parents: %s, set_gid: %s, sticky: %s)" % tup)
+        # set_gid and sticky bits are only set on new directories, so we need to determine the existing parent path
+        existing_parent_path = os.path.dirname(path)
+        try:
+            if parents:
+                # climb up until we hit an existing path or the empty string (for relative paths)
+                while existing_parent_path and not os.path.exists(existing_parent_path):
+                    existing_parent_path = os.path.dirname(existing_parent_path)
+                os.makedirs(path)
+            else:
+                os.mkdir(path)
+        except OSError, err:
+            _log.raiseException("Failed to create directory %s: %s" % (path, err))
+
+        # set group ID and sticky bits, if desired
+        bits = 0
+        if set_gid:
+            bits |= stat.S_ISGID
+        if sticky:
+            bits |= stat.S_ISVTX
+        if bits:
+            try:
+                new_subdir = path[len(existing_parent_path):].lstrip(os.path.sep)
+                new_path = os.path.join(existing_parent_path, new_subdir.split(os.path.sep)[0])
+                adjust_permissions(new_path, bits, add=True, relative=True, recursive=True, onlydirs=True)
+            except OSError, err:
+                _log.raiseException("Failed to set groud ID/sticky bit: %s" % err)
+    else:
+        _log.debug("Not creating existing path %s" % path)
+
+
+def read_file(path, log_error=True):
+    """Read contents of file at given path, in a robust way."""
+    f = None
+    # note: we can't use try-except-finally, because Python 2.4 doesn't support it as a single block
+    try:
+        f = open(path, 'r')
+        txt = f.read()
+        f.close()
+        return txt
+    except IOError, err:
+        # make sure file handle is always closed
+        if f is not None:
+            f.close()
+        if log_error:
+            _log.raiseException("Failed to read %s: %s" % (path, err))
+        else:
+            return None
```

Comment:
No need to `return None` when an exceptional case occurs, thrown an exception instead, that's what they're for. 
Flow control improves when using exceptions instead of having to check return values all the time.